### PR TITLE
FIX: first pass at keeping lazy components lazy

### PR DESCRIPTION
**Pull request by tacaswell**
Does not seem to be quite enough yet.

### CODE REVIEW

**Comment by tacaswell** [comment]
This seems to have cut the time of the test suite down by a factor of 2 so I guess that is good?  Need to have a think about how to test and bench mark this for real.

**Comment by tacaswell** [comment]
Master branch `============= 470 passed, 22 skipped, 6 warnings in 686.92 seconds =============`

This PR `======== 1 failed, 469 passed, 22 skipped, 6 warnings in 359.06 seconds ========`

So this has taken ~330s off the test suite run time, that seems like a success to me!

**Comment by tacaswell** [comment]
On my home machine, running just the pyepics version of the tests

current master: 311.89 seconds 
**master right before the kind merge: in 130.26 seconds**
gentle_gentle branch: 168.52 seconds (but 20 more tests than baseline)
**gentle_gentle branch: 126.75 seconds (not hints or kind test, 5 less tests that paseline)**

I think it is safe to say that this works.

File: ophyd/tests/conftest.py [hunk 1]
```diff
@@ -29,6 +29,11 @@ def __init__(self, pvname, form=None,
                  auto_monitor=True, enum_strs=None,
                  **kwargs):
 
+        # callbacks mechanism copied from pyepics
+        # ... but tweaked with a weakvaluedictionary so PV objects get
```
Comment (klauer) [inline ophyd/tests/conftest.py]: Looks like an outdated comment

**Comment by tacaswell** [comment]
```
________________________________________________________________________________________________________________ test_mca_read_attrs[pyepics] ________________________________________________________________________________________________________________

    @using_fake_epics_pv
    def test_mca_read_attrs():
        # default read_attrs
        mca = EpicsMCA(devs[0], name='test')
        default_normal_kind = ['preset_real_time', 'elapsed_real_time', 'spectrum']
>       assert set(default_normal_kind) == set(mca.read_attrs)
E       AssertionError: assert {'elapsed_rea...', 'spectrum'} == {'spectrum'}
E         Extra items in the left set:
E         'elapsed_real_time'
E         'preset_real_time'
E         Full diff:
E         - {'elapsed_real_time', 'preset_real_time', 'spectrum'}
E         + {'spectrum'}

ophyd/tests/test_mca.py:35: AssertionError
```

I can see this locally, but re-running it works.  Have only seen this with pyepics (not with caproto).  It is really weird that the fake pv has issues...

**Comment by tacaswell** [comment]
Please don't merge yet, this has a nasty mutating-shared-global-state bug.

**Comment by tacaswell** [comment]
This is ready to go.

**Comment by tacaswell** [comment]
both failures are caproto talking to real epics motor PVs.

**Comment by tacaswell** [comment]
The diff on this looks bigger than it really is because I moved the list proxy to be inside of Device, reviewing with white space changes suppressed may be helpful.

**Comment by danielballan** [comment]
All passed except the nightly build. Thanks, @tacaswell for indulging me with these new tests. I am now satisfied that we have not sacrificed the laziness feature accidentally in the course of adding `kind`.

**Pull request merged**